Products/CMFPlone/browser/search.py

```diff
@@ -183,7 +184,18 @@ def filter_types(self, types):
         plone_utils = getToolByName(self.context, "plone_utils")
         if not isinstance(types, list):
             types = [types]
-        return plone_utils.getUserFriendlyTypes(types)
+
+        # Get the vocabulary
+        vocab_factory = queryUtility(
+            IVocabularyFactory, "plone.app.vocabularies.ReallyUserFriendlyTypes"
+        )
+        vocab = vocab_factory(self.context)
+
+        # Filter types based on plone_utils and maintain vocabulary order
+        user_friendly_types = plone_utils.getUserFriendlyTypes(types)
+        sorted_types = [term.value for term in vocab if term.value in user_friendly_types]
+
+        return sorted_types
```

Comment:
I think the naming and all is that confusing, so a good amount of explanation is good for our future selfs.

```suggestion

        # We want to have the configured types to be exposed in the search sorted for humans (by translated Title).     
        # Those are stored in the Plone registry. They are called here UserFriendlyTypes.        
        # 
        # Confusingly, on the other hand we have the ReallyUserFriendlyTypes vocabulary from 
        # plone.app.vocabularies, which is already sorted accordingly and contain all possible types, 
        # except "Temp Folder" and "Plone Site". 

        # fetch the sorted ReallyUserFriendlyTypes vocabulary
          vocab_factory = queryUtility(
            IVocabularyFactory, "plone.app.vocabularies.ReallyUserFriendlyTypes"
        )
        vocab = vocab_factory(self.context)

        # get the configured values from the registry and pass the input types to be reduced to possible values
        user_friendly_types = plone_utils.getUserFriendlyTypes(types)

        # Filter the sorted ReallyUserFriendlyTypes down to the configured values from the registry,
        #  but keep the order.
        sorted_types = [term.value for term in vocab if term.value in user_friendly_types]

        return sorted_types
```